### PR TITLE
Use the variables defined in mounts (TPV destinations)

### PR DESCRIPTION
**Pull request by mira-miracoli**
Test for the new centralized [mounts repo](https://github.com/usegalaxy-eu/mounts), planned for Monday, 22nd May
Things to do before merge:
- [x] add `git clone git@github.com:usegalaxy-eu/mounts.git --branch main --single-branch --config core.sshCommand="ssh -o IdentitiesOnly=yes -i $JENKINS_SCRIPTS_PRIVATE_KEY" mounts` to the sn06 Jenkins Job
~- [ ] add `cd mounts && ansible-playbook -i [localhost,] all.yml && cd ..` to the sn06 Jenkins Job~
- [x] add a Makefile and add the make command to Jenkins

### CODE REVIEW

File: sn06.yml [hunk 1]
```diff
@@ -54,6 +54,7 @@
     - secret_group_vars/all.yml # All of the other assorted secrets...
     - secret_group_vars/keys.yml # SSH keys
     - templates/galaxy/config/job_conf.yml
+    - mounts/dest/all.yml
```
Comment (mira-miracoli) [inline sn06.yml:57]:
```suggestion
    - mounts/dest/all.yml
    - mounts/mountpoints.yml
```

File: files/galaxy/tpv/destinations.yml [hunk 2]
```diff
@@ -13,8 +13,7 @@ destinations:
       docker_net: bridge
       docker_auto_rm: true
       docker_set_user: ""
-      # $defaults expands to "$galaxy_root:ro,$tool_directory:ro,$job_directory:ro,$working_directory:rw,$default_file_path:rw"
-      docker_volumes: "$_CONDOR_SCRATCH_DIR:rw,$defaults,/cvmfs/data.galaxyproject.org:ro,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro,/data/dnb03/galaxy_db/:ro,/data/dnb05/galaxy_db/:ro,/data/dnb06/galaxy_db/:rw,/data/dnb07/galaxy_db/:rw,/data/dnb08/galaxy_db/:rw,/data/dp01/galaxy_db/:rw,/data/0/galaxy_db/:ro,/data/1/galaxy_db/:ro,/data/2/galaxy_db/:ro,/data/3/galaxy_db/:ro,/data/4/galaxy_db/:ro,/data/5/galaxy_import/galaxy_user_data/:ro,/data/6/galaxy_db/:ro,/data/7/galaxy_db/:ro"
+      docker_volumes: "$_CONDOR_SCRATCH_DIR:rw,$default,$job_directory:rw,$tool_directory:ro,{{ cvmfs.data.path }}:{{ cvmfs.data.docker_perm }}"
```
Comment (kysrpex) [inline files/galaxy/tpv/destinations.yml:16]: Is the change from `$defaults` to `$default` intended?

File: files/galaxy/tpv/destinations.yml [hunk 3]
```diff
@@ -28,8 +27,8 @@ destinations:
       submit_request_cpus: "{cores}"
       submit_request_memory: "{mem}G"
       singularity_enabled: true
-      singularity_volumes: "$_CONDOR_SCRATCH_DIR:rw,$job_directory:rw,$tool_directory:ro,$job_directory/outputs:rw,$working_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro,/data/dnb03/galaxy_db/:ro,/data/dnb05/galaxy_db/:ro,/data/dnb06/galaxy_db/:rw,/data/dnb07/galaxy_db/:rw,/data/dnb08/galaxy_db/:rw,/data/dp01/galaxy_db/:rw,/data/0/galaxy_db/:ro,/data/1/galaxy_db/:ro,/data/2/galaxy_db/:ro,/data/3/galaxy_db/:ro,/data/4/galaxy_db/:ro,/data/5/galaxy_import/galaxy_user_data/:ro,/data/6/galaxy_db/:ro,/data/7/galaxy_db/:ro,/usr/local/tools/:ro"
-      singularity_default_container_id: "/cvmfs/singularity.galaxyproject.org/all/centos:8.3.2011"
+      singularity_volumes: "$_CONDOR_SCRATCH_DIR:rw,$job_directory:rw,$tool_directory:ro,$job_directory/outputs:rw,$working_directory:rw,{% for vol in dnb.values() %}{{ vol.path }}/galaxy_db/:{{ vol.docker_perm }},{% endfor %}{{ tools.tools.path }}:{{ tools.tools.docker_perm }}"
```
Comment (kysrpex) [inline files/galaxy/tpv/destinations.yml:30]: If we switch to templates, then we need then the same trick required for setting up the TPV dry-run in Github actions (running Ansible just to fill the templates) to get the TPV linter to work again...

It's a pity that Jenkins does not do this honestly.
grafik

File: files/galaxy/tpv/destinations.yml [hunk 4]
```diff
@@ -101,17 +100,10 @@ destinations:
...
-      singularity_volumes: "$defaults,/cvmfs/data.galaxyproject.org:ro"
+      docker_volumes: "$default,$job_directory:rw,$tool_directory:ro,{{ cvmfs.data.path }}:{{ cvmfs.data.docker_perm }}"
     scheduling:
       require:
         - pulsar
-      accept:
-        - conda
-        - singularity
-        - docker
-        - condor-tpv
```
Comment (kysrpex) [inline files/galaxy/tpv/destinations.yml:114]: We may consider merging changes affecting scheduling tags in a different PR (or a least a specific commit just for this) to make things clearer. Of course at this point that involves interactive rebase and force push :wink:

**Comment by kysrpex** [comment]
Btw I had a look at the [mounts repo](https://github.com/usegalaxy-eu/mounts) and I like how you organized it :)

File: files/galaxy/tpv/destinations.yml [hunk 5]
```diff
@@ -101,17 +100,10 @@ destinations:
       dependency_resolution: "none"
       submit_request_cpus: "{cores}"
       submit_request_memory: "{mem}"
-      # $defaults expands to "$galaxy_root:ro,$tool_directory:ro,$job_directory:ro,$working_directory:rw,$default_file_path:rw"
-      docker_volumes: "$defaults,/cvmfs/data.galaxyproject.org:ro"
-      singularity_volumes: "$defaults,/cvmfs/data.galaxyproject.org:ro"
+      docker_volumes: "$default,$job_directory:rw,$tool_directory:ro,{{ cvmfs.data.path }}:{{ cvmfs.data.docker_perm }}"
```
Comment (mira-miracoli) [inline files/galaxy/tpv/destinations.yml:103]:
```suggestion
      docker_volumes: "$defaults,$job_directory:rw,$tool_directory:ro,{{ cvmfs.data.path }}:{{ cvmfs.data.docker_perm }}"
```

File: files/galaxy/tpv/destinations.yml (hunk 3 above)
Comment (bgruening) [reply files/galaxy/tpv/destinations.yml:30]: This way we define the string just once and reuse it everywhere.

Comment (kysrpex) [reply files/galaxy/tpv/destinations.yml:30]: But we are using stuff here (i.e. `for vol in dnb.values()`) that also needs to be shared with Ansible group variables, right? Defining a TPV variable somewhere may imply templating that file, and then we would run into the same problem.

Comment (kysrpex) [reply files/galaxy/tpv/destinations.yml:30]: For other use cases where things are only applied to TPV, I assume the answer is yes, at worst as an environment variable or similar.

However, I expect the solution to be more verbose than using Jinja templates (we could always construct that string with Jinja templates).

Comment (kysrpex) [reply files/galaxy/tpv/destinations.yml:30]: Assuming the TPV var would be reused a few times, that would work well :+1:

Comment (kysrpex) [reply files/galaxy/tpv/destinations.yml:30]: I am now wondering though: aren't we doing premature optimization here? Such a variable would only be reused here and in the basic docker destination right?

Comment (bgruening) [reply files/galaxy/tpv/destinations.yml:30]: Maybe, it was just a suggestion/Q.
I thought the volume string on its own, will be shared between all Docker and Singularity destinations etc ...

Comment (mira-miracoli) [reply files/galaxy/tpv/destinations.yml:30]: I think we can manage that with inheritance

Comment (kysrpex) [reply files/galaxy/tpv/destinations.yml:30]: > I think we can manage that with inheritance

Exactly I think this is already working well.

File: files/galaxy/tpv/destinations.yml.j2 [hunk 6]
```diff
@@ -101,9 +100,7 @@ destinations:
       dependency_resolution: "none"
       submit_request_cpus: "{cores}"
       submit_request_memory: "{mem}"
-      # $defaults expands to "$galaxy_root:ro,$tool_directory:ro,$job_directory:ro,$working_directory:rw,$default_file_path:rw"
-      docker_volumes: "$defaults,/cvmfs/data.galaxyproject.org:ro"
-      singularity_volumes: "$defaults,/cvmfs/data.galaxyproject.org:ro"
+      docker_volumes: "$defaults,$job_directory:rw,$tool_directory:ro,{{ cvmfs.data.path }}:{{ cvmfs.data.docker_perm }}"
```
Comment (kysrpex) [inline files/galaxy/tpv/destinations.yml.j2:103]: Same question as before, why do we need `$job_directory` to be writable now? Also `$tool_directory:ro` should be included in `$defaults`, shouldn't it?

File: files/galaxy/tpv/destinations.yml.j2 [hunk 7]
```diff
@@ -210,7 +207,7 @@ destinations:
         - type: singularity
           shell: "/bin/bash"
           resolve_dependencies: true
-          identifier: "/data/0/singularity_base_images/centos:8.3.2011"
+          identifier: "/data/{{ dnb.0.path }}/singularity_base_images/centos:8.3.2011"
```
Comment (kysrpex) [inline files/galaxy/tpv/destinations.yml.j2:210]:
```suggestion
          identifier: "{{ dnb.0.path }}/singularity_base_images/centos:8.3.2011"
```

[According to the mounts repo](https://github.com/usegalaxy-eu/mounts/blob/2c31cf88d6db95a6f6ae924f1680631d250948a1/mountpoints.yml#LL60C18-L60C18), `{{ dnb.0.path }}` corresponds to `/data/0`.

File: files/galaxy/tpv/destinations.yml.j2 [hunk 8]
```diff
@@ -13,8 +13,7 @@ destinations:
       docker_net: bridge
       docker_auto_rm: true
       docker_set_user: ""
-      # $defaults expands to "$galaxy_root:ro,$tool_directory:ro,$job_directory:ro,$working_directory:rw,$default_file_path:rw"
-      docker_volumes: "$_CONDOR_SCRATCH_DIR:rw,$defaults,/cvmfs/data.galaxyproject.org:ro,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro,/data/dnb03/galaxy_db/:ro,/data/dnb05/galaxy_db/:ro,/data/dnb06/galaxy_db/:rw,/data/dnb07/galaxy_db/:rw,/data/dnb08/galaxy_db/:rw,/data/dp01/galaxy_db/:rw,/data/0/galaxy_db/:ro,/data/1/galaxy_db/:ro,/data/2/galaxy_db/:ro,/data/3/galaxy_db/:ro,/data/4/galaxy_db/:ro,/data/5/galaxy_import/galaxy_user_data/:ro,/data/6/galaxy_db/:ro,/data/7/galaxy_db/:ro"
+      docker_volumes: "$_CONDOR_SCRATCH_DIR:rw,$default,$job_directory:rw,$tool_directory:ro,{{ cvmfs.data.path }}:{{ cvmfs.data.docker_perm }}"
```
[the remark this answers is not in the document]
Comment (mira-miracoli) [reply files/galaxy/tpv/destinations.yml.j2:16]: Is added now

File: files/galaxy/tpv/destinations.yml.j2 [hunk 9]
```diff
@@ -28,8 +27,8 @@ destinations:
       submit_request_cpus: "{cores}"
       submit_request_memory: "{mem}G"
       singularity_enabled: true
-      singularity_volumes: "$_CONDOR_SCRATCH_DIR:rw,$job_directory:rw,$tool_directory:ro,$job_directory/outputs:rw,$working_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro,/data/dnb03/galaxy_db/:ro,/data/dnb05/galaxy_db/:ro,/data/dnb06/galaxy_db/:rw,/data/dnb07/galaxy_db/:rw,/data/dnb08/galaxy_db/:rw,/data/dp01/galaxy_db/:rw,/data/0/galaxy_db/:ro,/data/1/galaxy_db/:ro,/data/2/galaxy_db/:ro,/data/3/galaxy_db/:ro,/data/4/galaxy_db/:ro,/data/5/galaxy_import/galaxy_user_data/:ro,/data/6/galaxy_db/:ro,/data/7/galaxy_db/:ro,/usr/local/tools/:ro"
-      singularity_default_container_id: "/cvmfs/singularity.galaxyproject.org/all/centos:8.3.2011"
+      singularity_volumes: "$_CONDOR_SCRATCH_DIR:rw,$job_directory:rw,$tool_directory:ro,$job_directory/outputs:rw,$working_directory:rw,{% for vol in dnb.values() %}{{ vol.path }}/galaxy_db/:{{ vol.docker_perm }},{% endfor %}{{ tools.tools.path }}:{{ tools.tools.docker_perm }}"
+      singularity_default_container_id: "{{ cvmfs.singularity.path }}/all/centos:8.3.2011"
```
[the remark this answers is not in the document]
Comment (mira-miracoli) [reply files/galaxy/tpv/destinations.yml.j2:31]: I fixed that

File: files/galaxy/tpv/destinations.yml.j2 (hunk 6 above)
Comment (mira-miracoli) [reply files/galaxy/tpv/destinations.yml.j2:103]: Should be identical with the string in infrastructure-playbooks now

**Comment by mira-miracoli** [comment]
Here is the current outcome:
https://gist.github.com/mira-miracoli/58db496dbaf8fc7dde5be993e1757bfb

**Comment by sanjaysrikakulam** [comment]
Looks good! Also, please include the vars file in other playbooks wherever necessary.

**Comment by kysrpex** [comment]
> Looks good! Also, please include the vars file in other playbooks wherever necessary.

Do you mind if we start with the TPV destinations and then do several PRs instead?

**Comment by kysrpex** [comment]
@sanjaysrikakulam After an offline discussion we are thinking of including the mounts diretory as a [Git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules) so that it is clear what other repository we are referencing and the reference is traceable (otherwise it is only evident if you look at Jenkins, which is private). What do you think?

**Comment by sanjaysrikakulam** [comment]
> @sanjaysrikakulam After an offline discussion we are thinking of including the mounts diretory as a [Git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules) so that it is clear what other repository we are referencing and the reference is traceable (otherwise it is only evident if you look at Jenkins, which is private). What do you think?

Sure! Otherwise, it could also simply be a `pre_task` that pulls the repo (respective template, and the file) and simplify the task and restrict it to be an Ansible task within the playbook. So everything is in one place.

**Comment by mira-miracoli** [comment]
I think the advantage of a submodule is, that it does not need to be added to every play is just a command that we run once and the mounts repo should be pulled automatically, whenever someone pulls the infrastructure-playbook repo

**Comment by bgruening** [comment]
That looks great!

Is the entire procedure documented in operations? Is it clear what our SOP is to add new mount points?

**Pull request merged**